### ISSUE TITLE
Implement better scheduler for Trino

### ISSUE THREAD

**Issue by sopel39** [trinodb/trino#10269]
Currently, Trino uses `AllAtOnceExecutionSchedule` execution policy by default. This makes all stages start at same time (even though splits for join build side are enumerated first). This means that join probe splits take valuable driver slots even though they cannot progress until build side stage completes. This can throttle cluster resource utilization for more complex queries or in high concurrency scenarios.

Trino already has `PhasedExecutionPolicy`, but it’s intended at minimizing memory/resource utilization rather than maximizing cpu/io utilization. However, it doesn’t schedule join probe side before join build side.

The scope of the ticket is to create a new scheduler that would:
* schedule join build source stages before join probe source staged
* schedule all intermediate stages
* schedule source stage if we know it’s data going to be consumed (e.g there is aggregation downstream)

Basically, improved all-at-once scheduler that doesn’t schedule source stage (to save driver slots) until it’s known that it’s data going to be consumed.

**Comment by sopel39** [comment]
cc @martint @dain

**Comment by findepi** [comment]
cc @losipiuk
